test/e2e/issue276_test.go

```diff
@@ -0,0 +1,242 @@
+		t.Error(err)
+	}
+
+	defer DumpJobsLogsOnError(t, framework.Global, namespace)
+	err = framework.AddToFrameworkScheme(apis.AddToScheme, &gitopsv1alpha1.GitOpsConfigList{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	eunomiaURI, found := os.LookupEnv("EUNOMIA_URI")
+	if !found {
+		eunomiaURI = "https://github.com/kohlstechnology/eunomia"
+	}
+	eunomiaRef, found := os.LookupEnv("EUNOMIA_REF")
+	if !found {
+		eunomiaRef = "master"
+	}
+
+	// Step 1: create a CR with a nonexistent TemplateSource ContextDir
+
+	const noDir = "test/e2e/testdata/no-directory"
+	gitops := &gitopsv1alpha1.GitOpsConfig{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "GitOpsConfig",
+			APIVersion: "eunomia.kohls.io/v1alpha1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "gitops-issue276",
+			Namespace: namespace,
+			Finalizers: []string{
+				"gitopsconfig.eunomia.kohls.io/finalizer",
+			},
+		},
+		Spec: gitopsv1alpha1.GitOpsConfigSpec{
+			TemplateSource: gitopsv1alpha1.GitConfig{
+				URI:        eunomiaURI,
+				Ref:        eunomiaRef,
+				ContextDir: noDir,
+			},
+			ParameterSource: gitopsv1alpha1.GitConfig{
+				URI:        eunomiaURI,
+				Ref:        eunomiaRef,
+				ContextDir: "test/e2e/testdata/empty-yaml",
+			},
+			Triggers: []gitopsv1alpha1.GitOpsTrigger{
+				{Type: "Change"},
+			},
+			TemplateProcessorImage: "quay.io/kohlstechnology/eunomia-base:dev",
+			ResourceHandlingMode:   "Apply",
+			ResourceDeletionMode:   "Delete",
+			ServiceAccountRef:      "eunomia-operator",
+		},
+	}
+	gitops.Annotations = map[string]string{"gitopsconfig.eunomia.kohls.io/initialized": "true"}
+
+	// Triple timeout.
+	// FIXME: as of writing this test, "backoffLimit" in job.yaml is set to 4,
+	// which means we need to wait until 5 Pod retries fail, eventually
+	// triggering a Job failure; the back-off time between the runs is
+	// unfortunately exponential and non-configurable, which makes this test
+	// awfully long. Try to at least make it possible to run in parallel with
+	// other tests.
+	err = framework.Global.Client.Create(context.TODO(), gitops, &framework.CleanupOptions{TestContext: ctx, Timeout: timeout * 3, RetryInterval: retryInterval})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Step 2: Wait until Job fails and check if it failed properly
+
+	err = wait.Poll(retryInterval, timeout, func() (done bool, err error) {
+		const name = "gitopsconfig-gitops-issue276-"
+		pod, err := GetPod(namespace, name, "quay.io/kohlstechnology/eunomia-base:dev", framework.Global.KubeClient)
+		switch {
+		case apierrors.IsNotFound(err):
+			t.Logf("Waiting for availability of %s pod", name)
+			return false, nil
+		case err != nil:
+			return false, err
+		case pod != nil && pod.Status.Phase == "Failed":
+			logs, err := GetPodLogs(pod, framework.Global.KubeClient)
+			if err != nil {
+				t.Fatal(err)
+			}
+			if !strings.Contains(logs, fmt.Sprintf("ERROR - directory %s does not exist in the remote repository", noDir)) {
+				t.Fatal(err)
```

Comment:
err will be nil here, no? should we print some more descriptive info here? did you verify if the test fails when you revert the code to the version before fix?

Reply:
I'll add a description here.
The test fails with the previous code because the Poll function reaches timeout at some point and an error is set:
https://github.com/kubernetes/apimachinery/blob/731dcecc205498f52a21b12e311af095efb4b188/pkg/util/wait/wait.go#L421-L422
(the WaitFor function is used by pollInternal which is used by Poll)